### PR TITLE
[DependencyInjection] add `Autowire` parameter attribute

### PR DESCRIPTION
**Pull request by kbond**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | todo

Replaces #45573 & #44780 with a single new `Autowire` attribute:

```php
class MyService
{
    public function __construct(
        #[Autowire(service: 'some_service')]
        private $service1,

        #[Autowire(expression: 'service("App\\Mail\\MailerConfiguration").getMailerMethod()')
        private $service2,

        #[Autowire(value: '%env(json:file:resolve:AUTH_FILE)%')]
        private $parameter1,

        #[Autowire(value: '%kernel.project_dir%/config/dir')]
        private $parameter2,
    ) {}
}
```

Works with controller arguments as well:

```php
class MyController
{
    public function someAction(
        #[Autowire(service: 'some_service')]
        $service1,

        #[Autowire(expression: 'service("App\\Mail\\MailerConfiguration").getMailerMethod()')
        $service2,

        #[Autowire(value: '%env(json:file:resolve:AUTH_FILE)%')]
        $parameter1,

        #[Autowire(value: '%kernel.project_dir%/config/dir')]
        $parameter2,
    ): Response {}
}
```

### CODE REVIEW

**Review by nicolas-grekas — COMMENTED** [review]
~What about calling `Yaml::parse()` before calling `YamlFileLoader::resolveServices()`?~

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -619,4 +629,20 @@ private function populateCombinedAliases(ContainerBuilder $container): void
...
         }
     }
+
+    private static function resolveAutowireAttribute(\ReflectionAttribute $attribute, \ReflectionClass $class): mixed
+    {
+        if (self::$autowireAttributeResolver) {
+            return (self::$autowireAttributeResolver)($attribute, $class);
+        }
+
+        self::$autowireAttributeResolver = function (\ReflectionAttribute $attribute, \ReflectionClass $class) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: the previous "if" can be removed and this be changed to `self::$autowireAttributeResolver ??=`

**Comment by nicolas-grekas** [comment]
> What about calling Yaml::parse() before calling YamlFileLoader::resolveServices()?

I'm withdrawing this comment, bad idea :)

Another idea, using named arguments:

- `#[Autowire(service: MailerConfiguration::class)`
- `#[Autowire(expr: 'service("App\Mail\MailerConfiguration").getMailerMethod()')`
- `#[Autowire(value: '%kernel.project_dir%'/var/foo')`

I think that's my preference: no custom yaml-like strings and more explicit than standalone attributes.

**Comment by kbond** [comment]
I've made this change and also added the _raw_ `parameter` argument (updated PR description).

What about controller action arguments? Separate PR?

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+{
+    public readonly string $value;
+
+    /**
+     * Use only ONE of the following.
+     *
+     * @param string|null $service    Service ID (ie "some.service")
+     * @param string|null $expression Expression (ie 'service("some.service").someMethod()')
+     * @param string|null $parameter  Parameter name (ie "kernel.environment")
+     * @param string|null $value      Any valid YAML service config (ie "%kernel.project_dir%/some/path")
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: this is unrelated to yaml

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public function __construct(
+        ?string $service = null,
+        ?string $expression = null,
+        ?string $parameter = null,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: I'm not sure adding this one is needed; value is enough to me. DI parameters are very often referenced using `%...%` so it's going to be quite natural to continue this way IMHO.

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        ?string $value = null
+    ) {
+        if ($service) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: We might want to add a guard clause:

`if (!($service xor $expression xor $value)) { throw LogicException( Can only set one of ... ))`

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 5]
```diff
@@ -619,4 +629,16 @@ private function populateCombinedAliases(ContainerBuilder $container): void
             }
         }
     }
+
+    private static function resolveAutowireAttribute(\ReflectionAttribute $attribute, \ReflectionClass $class): mixed
+    {
+        self::$autowireAttributeResolver ??= function (\ReflectionAttribute $attribute, \ReflectionClass $class) {
+            $parseDefinitions = new \ReflectionMethod(YamlFileLoader::class, 'resolveServices');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: better not use YamlFileLoader now that we have the attribute already parsed by php

**Review by nicolas-grekas — COMMENTED** [review]
> What about controller action arguments? Separate PR?

same PR to me, it's the same topic

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            return;
+        }
+
+        if ($value) {
+            $this->value = $value;
+
+            return;
+        }
+
+        throw new LogicException('One of $service, $expression, $value must be set.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: isn't this case already covered by the xor above?

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+     * @param string|null $service    Service ID (ie "some.service")
+     * @param string|null $expression Expression (ie 'service("some.service").someMethod()')
+     * @param string|null $value      Parameter value (ie "%kernel.project_dir%/some/path")
+     */
+    public function __construct(
+        ?string $service = null,
+        ?string $expression = null,
+        ?string $value = null
+    ) {
+        if (!($service xor $expression xor $value)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: null !== $value, to still allow $value = '0'

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            throw new LogicException('Can only set one of $service, $expression, $value.');
+        }
+
+        if ($service) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: let's use a "match" instead of those "if"s?

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            $this->value = new Reference($service);
+
+            return;
+        }
+
+        if ($expression) {
+            $this->value = new Expression($expression);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: `throw new LogicException('Unable to use expressions as the Symfony ExpressionLanguage component is not installed. Try running "composer require symfony/expression-language".');`

if the class doesn't exist

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+     * @param string|null $expression Expression (ie 'service("some.service").someMethod()')
+     * @param string|null $value      Parameter value (ie "%kernel.project_dir%/some/path")
+     */
+    public function __construct(
+        ?string $service = null,
+        ?string $expression = null,
+        ?string $value = null
+    ) {
+        if (!($service xor $expression xor $value)) {
+            throw new LogicException('Can only set one of $service, $expression, $value.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: What about `#[Autowire] attribute must declare exactly one of $service, $expression, or $value.`?

**Review by nicolas-grekas — COMMENTED** [review]
Here are some minor comments. I like it. How about you? Let's close the two previous ones?

**Comment by kbond** [comment]
> Here are some minor comments. I like it. How about you? Let's close the two previous ones?

Made these changes. I am happy with it as well - let's close them!

Few questions:
1. Should I change the service Reference behaviour for nullable parameters?
2. For controller arguments: are only service's allowed? No expressions/parameters?

**Comment by nicolas-grekas** [comment]
> Should I change the service Reference behaviour for nullable parameters?

That'd make sense yes

> For controller arguments: are only service's allowed? No expressions/parameters?

expressions/parameters should be allowed. There is a hack using 'current' as factory right now, you'll see it in the pass.

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        ?string $value = null
+    ) {
+        if (!($service xor $expression xor null !== $value)) {
+            throw new LogicException('#[Autowire] attribute must declare exactly one of $service, $expression, or $value.');
+        }
+
+        $this->value = match (true) {
+            null !== $service => new Reference($service),
+            null !== $expression => class_exists(Expression::class) ? new Expression($expression) : throw new LogicException('Unable to use expressions as the Symfony ExpressionLanguage component is not installed. Try running "composer require symfony/expression-language".'),
+            default => $value,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: `null !== $value`, not default, just to be strict

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\ExpressionLanguage\Expression;
+
+/**
+ * Attribute to tell a parameter how to be autowired.
+ *
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+#[\Attribute(\Attribute::TARGET_PARAMETER)]
+final class Autowire
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php]: I don't think that `final` provides us anything, so we can remove it
same in the test case as always ;)

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes_80.php [hunk 13]
```diff
@@ -26,3 +27,18 @@ class AutowireProperty
...
+    public function __construct(
+        #[Autowire(service: 'some.id')]
+        public \stdClass $service,
+        #[Autowire(expression: "parameter('some.parameter')")]
+        public string $expression,
+        #[Autowire(value: '%some.parameter%/bar')]
+        public ?string $value,
+        #[Autowire(service: 'invalid.id')]
+        public ?\stdClass $invalid = null,
+    ){
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes_80.php]: missing space between brackets

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+     * Use only ONE of the following.
+     *
+     * @param string|null $service    Service ID (ie "some.service")
+     * @param string|null $expression Expression (ie 'service("some.service").someMethod()')
+     * @param string|null $value      Parameter value (ie "%kernel.project_dir%/some/path")
+     */
+    public function __construct(
+        ?string $service = null,
+        ?string $expression = null,
+        ?string $value = null
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: which one should be first?
what do we want `#[Autowire('foo')]` to mean?

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: I think `$service` would be the most common use. Wish we could use named constructors with attributes :)

**Comment by kbond** [comment]
Ok, I have the controller argument support working but implementation might need some work.

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php [hunk 15]
```diff
@@ -120,6 +121,21 @@ public function process(ContainerBuilder $container)
                 $args = [];
                 foreach ($parameters as $p) {
                     /** @var \ReflectionParameter $p */
+                    if (class_exists(Autowire::class) && $attr = $p->getAttributes(Autowire::class)[0] ?? null) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: reading attributes should happen last: precedence should go from most explicit to less explicit, and most explicit is always hand written config (attributes on the DI tag, etc)

Comment (derrabus) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: Is the `class_exists()` check necessary? Calling `$p->getAttributes(InvalidClass::class)` should be a safe operation.

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php [hunk 16]
```diff
@@ -120,6 +121,21 @@ public function process(ContainerBuilder $container)
                 $args = [];
                 foreach ($parameters as $p) {
                     /** @var \ReflectionParameter $p */
+                    if ($attr = $p->getAttributes(Autowire::class)[0] ?? null) {
+                        $value = $attr->newInstance()->value;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: class_exists was needed because we create an instance here ;)

Comment (kbond) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: Doesn't the above if statement evaluate to false if `Autowire` doesn't exist?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: Nope, attributes being declarative, their implantation is not required until one calls newInstance

Comment (kbond) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: TIL :)

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php (hunk 15 above)
Comment (kbond) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: I'm... a bit lost here.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: this if should go just before the one on L164/180 that reads as `if ($type &&...`

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php [hunk 17]
```diff
@@ -120,6 +121,21 @@ public function process(ContainerBuilder $container)
                 $args = [];
                 foreach ($parameters as $p) {
                     /** @var \ReflectionParameter $p */
+                    if (class_exists(Autowire::class) && $attr = $p->getAttributes(Autowire::class)[0] ?? null) {
+                        $value = $attr->newInstance()->value;
+
+                        if ($value instanceof Reference) {
+                            $args[$p->name] = new Reference($value, $p->allowsNull() ? ContainerInterface::NULL_ON_INVALID_REFERENCE : ContainerInterface::RUNTIME_EXCEPTION_ON_INVALID_REFERENCE);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: the "else" part should be `$value->getInvalidBehavior()`

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php (hunk 15 above)
Comment (kbond) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: Thanks :)

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php (hunk 17 above)
Comment (kbond) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: After moving, I was able to use `$invalidBehavior`.

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php [hunk 18]
```diff
@@ -148,7 +149,7 @@ public function process(ContainerBuilder $container)
                         }
 
                         continue;
-                    } elseif (!$type || !$autowire || '\\' !== $target[0]) {
+                    } elseif ((!$type || !$autowire || '\\' !== $target[0]) && null === $p->getAttributes(Autowire::class) ?? null) {
```
Comment (kbond) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: Note: I had to add this check to skip Autowire parameters/expressions.

File: src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php [hunk 19]
```diff
@@ -434,6 +435,36 @@ public function testBindWithTarget()
...
+        $container->getDefinition($locatorId)->setPublic(true);
+
+        $container->compile();
+
+        $locator = $container->get($locatorId);
+
+        $this->assertInstanceOf(\stdClass::class, $locator->get('foo::fooAction')->get('service1'));
+        $this->assertSame('foo/bar', $locator->get('foo::fooAction')->get('value'));
+        $this->assertSame('foo', $locator->get('foo::fooAction')->get('expression'));
+        $this->assertNull($locator->get('foo::fooAction')->get('service2'));
```
Comment (kbond) [inline src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php]: This test now fails (it was passing previously). Should I change to:

```suggestion
        $this->assertFalse($locator->get('foo::fooAction')->has('service2'));
```

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php (hunk 18 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: `} elseif (!$autowire || (!($attr = $p->getAttributes(Autowire::class)) && (!$type || '\\' !== $target[0])) {`

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php [hunk 20]
```diff
@@ -161,6 +162,21 @@ public function process(ContainerBuilder $container)
                         continue;
                     }
 
+                    if (class_exists(Autowire::class) && $attr = $p->getAttributes(Autowire::class)[0] ?? null) {
+                        $value = $attr->newInstance()->value;
+
+                        if ($value instanceof Reference) {
+                            $args[$p->name] = new Reference($value, $invalidBehavior);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: `$args[$p->name] = $type ? new TypedReference($value, $type, $invalidBehavior, $p->name) : new Reference($value, $invalidBehavior);`

File: src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php (hunk 19 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php]: only if it makes sense! I made some suggestions because the current logic is flawed, maybe that's going fix this?

Comment (kbond) [reply src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php]: I've made the logic changes but this assertion still fails. I don't know enough about how the method is invoked by this system. Should this actually return null or does the system skip if the container doesn't have the service and the arg is nullable?

Comment (kbond) [reply src/Symfony/Component/HttpKernel/Tests/DependencyInjection/RegisterControllerArgumentLocatorsPassTest.php]: Discovered it's expected for nullable arguments to not be in the service locator.

**Review by nicolas-grekas — APPROVED** [review]
I like it, thanks!
(I pushed [some tweaks](https://github.com/symfony/symfony/compare/d9b0417db565847501ebe37964475d0788a61576..d43fe42109))

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php (hunk 14 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: Now, I'm wondering: what about this implementation?
```php
class Autowire
{
    public readonly string|Expression|Reference $value;

    public function __construct(
        ?string $value = null,
        ?string $service = null,
        ?string $expression = null,
    ) {
        [...]

        if (null !== $value && str_starts_with($value, '@')) {
            match (true) {
                str_starts_with($value, '@@') => $value = substr($value, 1),
                str_starts_with($value, '@=') => $expression = substr($value, 2),
                default => $service = $value,
            };
        }

        [...]
}
```

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: Can you show some usage examples?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: `#[Autowire('@some_service')]`
but also
`#[Autowire('%kernel.debug%')]`

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: Oh ok, why the `@@` check?

And also:

`#[Autowire('@=service("App\\Mail\\MailerConfiguration").getMailerMethod()')]`

But then, this would not work, correct?

`#[Autowire(SomeService::class)]`

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: for `@@`, see https://github.com/symfony/symfony/blob/c1aa7b5e98d1fbc68c4594cc3f04ae2ec03c74ec/src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php#L835

> But then, this would not work, correct?

correct, this would:
`#[Autowire('@'.SomeService::class)]`

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: and this also of course:
`#[Autowire(service: SomeService::class)]`

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: Got it, since we're not using the Yaml loader, we'll have to duplicate the logic to parse these? What about `@!`/`@?`, should this syntax be supported as well? _I might be missing some nuance/context here._

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: Sorry, I now see `@@` is just a way to escape raw values that start with `@`.

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: > `default => $service = $value,`

Shouldn't this remove the `@`?

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: `@?` is handled by the _nullability_ of the parameter. Should we handle the `ContainerInterface::IGNORE_ON_UNINITIALIZED_REFERENCE` (`@!`) case?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: > Shouldn't this remove the @?
 yes!

> Should we handle the `@*`

no need to imho

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: Ok, I'm happy with this syntax! Want me to update the PR to support?

**Review by GromNaN — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php (hunk 14 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:38]: yes please!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @kbond.

File: src/Symfony/Component/DependencyInjection/Attribute/Autowire.php [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+use Symfony\Component\DependencyInjection\Exception\LogicException;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\ExpressionLanguage\Expression;
+
+/**
+ * Attribute to tell a parameter how to be autowired.
+ *
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+#[\Attribute(\Attribute::TARGET_PARAMETER)]
```
Comment (qdequippe) [inline src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: Do you think this `#[\Attribute(\Attribute::TARGET_PARAMETER, \Attribute::TARGET_PROPERTY)]` can be used here? When using Attribute with promoted property constructor.

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: `#[\Attribute(\Attribute::TARGET_PARAMETER]` can be used on promoted properties. See https://github.com/symfony/symfony/pull/45657/files#diff-a4f7e7528218aa6d151b1e68834393c6ae958540ec31ffa49ce3207998e77dd2R33

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: why would we? `TARGET_PARAMETER` is enough to target a promoted argument

Comment (qdequippe) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: Yes you're right `TARGET_PARAMETER` is enough here. Thank you for your quick response :)

Comment (michaljusiega) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: I'm facing the same problem. 

`TARGET_PARAMETER` is working as well with properties for constructor, however not working especially with reflection mechanism on reflection property.

See https://3v4l.org/gDeYA what I mean.

Also PHPStorm suggests not to use this attribute on property promoted:
obraz

cc @kbond @nicolas-grekas @qdequippe

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: Could this be a bug in PHP or a missing detail in the implementation? The opposite is a problem too: https://3v4l.org/tBjj2. Somewhat related PHPStan issue: https://github.com/phpstan/phpstan/issues/4418.

Comment (michaljusiega) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: As far as I understand, PHPStan literally "silence"  this situation, right?

For me isn't looks like a PHP bug, in my opinion we should add TARGET_PROPERTY flag. ;)

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: To me it's a limitation of the engine and the engine should be improved: from a descriptive pov, both variants give all the possible hints to make things clear. The engine is not clever enough yet.
The alternative is to not use CPP but it would be a shame to accept this state of things :) Patching the attribute would be wrong: Autowire should NOT be allowed on properties.
Could you please open a bug report about this on php-src?

Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: > See https://3v4l.org/gDeYA what I mean.

Calling `newInstance()` for arbitrary attributes is a footgun. Attributes could point to undefined classes and the attribute constructor could have side effects. The reflection API by design provides you with enough information to decide whether you need an actual instance of an attribute. Only call `newInstance()` on attributes you know and that your code actually needs to work with.

I agree with @nicolas-grekas that there's nothing we should do on our side.

Comment (kbond) [reply src/Symfony/Component/DependencyInjection/Attribute/Autowire.php:23]: I agree as well. At least until a time where the attribute can actually be used with [property injection](https://symfony.com/doc/current/service_container/injection_types.html#property-injection).
